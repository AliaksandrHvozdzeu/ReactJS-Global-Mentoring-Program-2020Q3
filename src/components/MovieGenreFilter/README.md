#Movie filter by genre component
