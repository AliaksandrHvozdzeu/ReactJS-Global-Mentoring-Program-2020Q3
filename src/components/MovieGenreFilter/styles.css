.genre-filter-item button {
  text-transform: uppercase;
}

.genre-filter-item:not(:last-child) {
  margin-right: 2rem;
}

.genre-filter-item, .genre-filter-item button {
  list-style-type: none;
  text-decoration: none;
  color: var(--white);
  position: relative;
  background-color: transparent;
  font-size: 1.2rem;
  outline: none;
  line-height: 4rem;
}

.genre-filter-item button:hover::after, .genre-filter-item button.active::after {
  position: absolute;
  content: '';
  border-bottom: 3px solid var(--red);
  left: 0;
  right: 0;
  bottom: 0;
  z-index: 2;
  border-radius: 1px;
}
