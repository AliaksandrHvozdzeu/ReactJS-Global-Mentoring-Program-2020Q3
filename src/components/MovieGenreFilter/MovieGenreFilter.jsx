import React, { useState } from 'react';
import PropTypes from 'prop-types';
import './styles.css';

export default function MovieGenreFilter({ genres, action, active }) {

  const [activeGenre, setActiveGenre] = useState(active || genres[0]);

  const onClick = (genre) => {
    action(genre);
    setActiveGenre(genre);
  };

  return (
    <>
      <ul className="genre-filter">
        {genres.map((genre) => (
          <li key={genre} className="genre-filter-item">
            <button type="button" onClick={() => onClick(genre)} className={(genre === activeGenre ? 'active' : '')}>
              {genre}
            </button>
          </li>
        ))}
      </ul>
    </>
  );

}

MovieGenreFilter.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  action: PropTypes.func.isRequired,
  active: PropTypes.string,
};

MovieGenreFilter.defaultProps = {
  active: null,
};
