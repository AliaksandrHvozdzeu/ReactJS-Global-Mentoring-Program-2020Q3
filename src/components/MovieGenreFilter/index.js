import MovieGenreFilter from './js/MovieGenreFilter';

export default MovieGenreFilter;
