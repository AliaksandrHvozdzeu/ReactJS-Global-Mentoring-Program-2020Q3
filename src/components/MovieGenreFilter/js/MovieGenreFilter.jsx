import React, { useCallback, useState } from 'react';
import PropTypes from 'prop-types';
import Constants from '../../../constants';
import '../css/MovieGenreFilter.css';

export default function MovieGenreFilter({ genres, action, active }) {

  const [activeGenre, setActiveGenre] = useState(active || genres[0]);

  const onClick = (genre) => {
    action(genre);
    setActiveGenre(genre);
  };

  const onSelect = useCallback(
    (event) => {
      const value = event.target;
      action(value);
      setActiveGenre(value);
    },
    [action, setActiveGenre],
  );

  const className = useCallback(
    (genre) => (genre === activeGenre ? 'active' : ''),
    [activeGenre],
  );

  const select = (
    <select
      id="genre-filter"
      className="genre-filter"
      onChange={onSelect}
      value={activeGenre}
    >
      {genres.map((option) => (
        <option key={option} className="genre-filter-item" value={option}>
          {option}
        </option>
      ))}
    </select>
  );

  const list = (
    <ul className="genre-filter">
      {genres.map((genre) => (
        <li key={genre} className="genre-filter-item">
          <button
            type="button"
            onClick={() => onClick(genre)}
            className={className(genre)}
          >
            {genre}
          </button>
        </li>
      ))}
    </ul>
  );

  return (
    <>
      {genres.length <= Constants.DEFAULT_COUNT_FILTER_ITEM && list}
      {genres.length > Constants.DEFAULT_COUNT_FILTER_ITEM && select}
    </>
  );

}

MovieGenreFilter.propTypes = {
  genres: PropTypes.arrayOf(PropTypes.string).isRequired,
  action: PropTypes.func.isRequired,
  active: PropTypes.string,
};

MovieGenreFilter.defaultProps = {
  active: null,
};
