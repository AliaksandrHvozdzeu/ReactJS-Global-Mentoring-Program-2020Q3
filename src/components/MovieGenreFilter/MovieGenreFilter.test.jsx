import React from "react";
import rerender from "react-test-renderer";
import MovieGenreFilter from "./js/MovieGenreFilter";
import { shallow } from "enzyme";
import { jest } from "@jest/globals";

describe("when MovieGenreFilter", () => {

  const action = jest.fn();
  const active = "";
  const genresList = ["Comedy", "Documentary", "Horror", "Crime", "Drama", "Romance", "Adventure", "Comedy"];

  test("then snapshot created", () => {
    const component = rerender.create(<MovieGenreFilter genres={[]}
                                                        action={action}
                                                        active={active} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("genres select", () => {
    const modal = shallow(<MovieGenreFilter genres={genresList}
                                            action={action}
                                            active={active} />);
    expect(modal.find(".genre-filter")).toBeTruthy();
  });

  test("genres list", () => {
    const modal = shallow(<MovieGenreFilter genres={genresList}
                                            action={action}
                                            active={active} />);
    expect(modal.find(".genre-filter")).toBeTruthy();
  });

  test('onChange', () => {
    const component = shallow(<MovieGenreFilter genres={genresList}
                                                action={action}
                                                active={active} />);
    const selector = component.find('#genre-filter');
    const event = {
      preventDefault() {},
      target: { value: genresList[0] }
    };
    selector.simulate('change', event.target);
    const elem = component.find('.genre-filter-item');
    expect(elem.length).toBe(8);
  });

  test('onClick', () => {
    const component = shallow(<MovieGenreFilter genres={genresList}
                                                action={action}
                                                active={active} />);
    const selector = component.find('#genre-filter');
    selector.simulate('click', genresList[0]);
    const elem = component.find('.genre-filter-item');
    expect(elem.length).toBe(8);
  });

});
