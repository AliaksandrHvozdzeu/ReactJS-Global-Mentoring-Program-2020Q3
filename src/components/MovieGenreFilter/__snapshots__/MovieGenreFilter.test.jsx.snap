// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when MovieGenreFilter then snapshot created 1`] = `
<ul
  className="genre-filter"
/>
`;
